lms/djangoapps/badges/models.py

```diff
@@ -0,0 +1,183 @@
+from django.contrib.auth.models import User
+from django.core.exceptions import ValidationError
+from django.db import models
+from django.utils.translation import ugettext_lazy as _
+from lazy import lazy
+
+from xmodule_django.models import CourseKeyField
+from jsonfield import JSONField
+
+
+def validate_badge_image(image):
+    """
+    Validates that a particular image is small enough, of the right type, and square to be a badge.
+    """
+    if image.width != image.height:
+        raise ValidationError(_(u"The badge image must be square."))
+    if not image.size < (250 * 1024):
+        raise ValidationError(_(u"The badge image file size must be less than 250KB."))
+
+
+class BadgeClass(models.Model):
+    """
+    Specifies a badge class to be registered with a backend.
+    """
+    slug = models.SlugField(max_length=255)
+    issuing_component = models.SlugField(max_length=50, default='', blank=True)
+    display_name = models.CharField(max_length=255)
+    course_id = CourseKeyField(max_length=255, blank=True, default=None)
+    description = models.TextField()
+    criteria = models.TextField()
+    # Mode a badge was awarded for. Included for legacy/migration purposes.
+    mode = models.CharField(max_length=100, default='', blank=True)
+    image = models.ImageField(upload_to='badge_classes', validators=[validate_badge_image])
+
+    def __unicode__(self):
+        return u"<Badge '{slug}' for '{issuing_component}'>".format(
+            slug=self.slug, issuing_component=self.issuing_component
+        )
+
+    @classmethod
+    def get_badge_class(
+            cls, slug, issuing_component, display_name, description, criteria, image_file_handle,
+            mode='', course_id=None, create=True
+    ):
+        """
+        Looks up a badge class by its slug, issuing component, and course_id and returns it should it exist.
+        If it does not exist, and create is True, creates it according to the arguments. Otherwise, returns None.
+        """
+        if not course_id:
+            course_id = CourseKeyField.Empty
+        try:
+            return cls.objects.get(slug=slug, issuing_component=issuing_component, course_id=course_id)
+        except cls.DoesNotExist:
+            if not create:
+                return None
+        badge_class = cls(
+            slug=slug,
+            issuing_component=issuing_component,
+            display_name=display_name,
+            course_id=course_id,
+            mode=mode,
+            description=description,
+            criteria=criteria,
+        )
+        badge_class.image.save(image_file_handle.name, image_file_handle)
+        badge_class.full_clean()
+        badge_class.save()
+        return badge_class
+
+    @lazy
+    def backend(self):
+        """
+        Loads the badging backend.
+        """
+        module, klass = settings.BADGING_BACKEND.rsplit('.', 1)
+        module = import_module(module)
+        return getattr(module, klass)()
```

Comment:
Perhaps we already have legacy settings that force our hand, but: instead of settings.BADGR_URL, settings.BADGR_API_TOKEN, and settings.BADGR_ISSUER_SLUG, wouldn't it be better to have a settings.BADGING_SETTINGS set to a dict with whatever settings the backend needs, including the BACKEND to use?  This mimics the DATABASE settings, and now the TEMPLATE settings.  Then here you could pass the BADGING_SETTINGS to the constructor.  This would keep the settings cleaner.